src/ripple/basics/scope.h

```diff
@@ -0,0 +1,196 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2021 Ripple Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_SCOPE_H_INCLUDED
+#define RIPPLE_BASICS_SCOPE_H_INCLUDED
+
+#include <exception>
+#include <type_traits>
+#include <utility>
+
+namespace ripple {
+
+// RAII scope helpers.  As specified in Library Fundamental, Version 3
+// Basic design of idea:  https://www.youtube.com/watch?v=WjTrfoiB0MQ
+// Specification:
+//   http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4873.html#scopeguard
+
+template <class EF>
+class scope_exit
+{
+    EF exit_function_;
+    bool execute_on_destruction_{true};
+
+public:
+    ~scope_exit()
+    {
+        if (execute_on_destruction_)
+            exit_function_();
+    }
+
+    scope_exit(scope_exit&& rhs) noexcept(
+        std::is_nothrow_move_constructible_v<EF> ||
+        std::is_nothrow_copy_constructible_v<EF>)
+        : exit_function_{std::forward<EF>(rhs.exit_function_)}
+        , execute_on_destruction_{rhs.execute_on_destruction_}
+    {
+        rhs.release();
+    }
+
+    scope_exit&
+    operator=(scope_exit&&) = delete;
+
+    template <class EFP>
+    explicit scope_exit(
+        EFP&& f,
+        std::enable_if_t<
+            !std::is_same_v<std::remove_cv_t<EFP>, scope_exit> &&
+            std::is_constructible_v<EF, EFP>>* =
+            0) noexcept(std::is_nothrow_constructible_v<EF, EFP> || std::is_nothrow_constructible_v<EF, EFP&>)
```

Comment:
It's a shame the `enable_if` has to be a parameter - which hides it away from where we see it in the rest of the rippled code. But I think we're forced to do that because this is a constructor.

Reply:
Right.  And `requires` isn't in our toolbox yet.  (`requires` is going to be great...)